src/Compiler/CompileExpr.idr

```diff
@@ -165,63 +165,99 @@ magic ms e = go ms e where
 natMinus : FC -> FC -> forall vars. Vect 2 (CExp vars) -> CExp vars
 natMinus fc fc' [m,n] = CApp fc (CRef fc' (UN "prim__sub_Integer")) [m, n]
 
-natHack : CExp vars -> CExp vars
-natHack = magic
-    [ MagicCCon typesNS "Z" 0
-         (\ fc, [] => CPrimVal fc (BI 0))
-    , MagicCCon typesNS "S" 1
-         (\ fc, [k] => CApp fc (CRef fc (UN "prim__add_Integer")) [CPrimVal fc (BI 1), k])
-    , MagicCRef typesNS "natToInteger" 1
+-- TODO: next release remove this and use %builtin pragma
+natHack : List Magic
+natHack =
+    [ MagicCRef (NS typesNS (UN "natToInteger")) 1
          (\ _, _, [k] => k)
-    , MagicCRef typesNS "integerToNat" 1
+    , MagicCRef (NS typesNS (UN "integerToNat")) 1
          (\ fc, fc', [k] => CApp fc (CRef fc' (NS typesNS (UN "prim__integerToNat"))) [k])
-    , MagicCRef typesNS "plus" 2
+    , MagicCRef (NS typesNS (UN "plus")) 2
          (\ fc, fc', [m,n] => CApp fc (CRef fc' (UN "prim__add_Integer")) [m, n])
-    , MagicCRef typesNS "mult" 2
+    , MagicCRef (NS typesNS (UN "mult")) 2
          (\ fc, fc', [m,n] => CApp fc (CRef fc' (UN "prim__mul_Integer")) [m, n])
-    , MagicCRef natNS "minus" 2 natMinus
+    , MagicCRef (NS natNS (UN "minus")) 2 natMinus
     ]
 
-
-isNatCon : Name -> Bool
-isNatCon (NS ns (UN n))
-   = (n == "Z" || n == "S") && ns == typesNS
-isNatCon _ = False
-
-natBranch : CConAlt vars -> Bool
-natBranch (MkConAlt n _ _ _) = isNatCon n
-
-trySBranch : CExp vars -> CConAlt vars -> Maybe (CExp vars)
-trySBranch n (MkConAlt (NS ns (UN nm)) _ [arg] sc)
-  = do guard (nm == "S" && ns == typesNS)
+-- get all transformation from %builtin pragmas
+builtinMagic : Ref Ctxt Defs => Core (forall vars. CExp vars -> CExp vars)
+builtinMagic = do
+    defs <- get Ctxt
+    let b = defs.builtinTransforms
+    let nats = concat $ uncurry builtinMagicNat <$> values (natTyNames b)
+    pure $ magic $ natHack ++ nats
+  where
+    builtinMagicNat : Name -> Name -> List Magic
```

Comment:
Can we add a record type `NatBuiltin` that stores these names? I am uneasy about having
multiple arguments with the same type and would rather the function relied on field names
than the arguments being passed in the right order.

Reply:
Similarly I'd rather have `NameMap SUC` and `NameMap ZERO` than `NameMap ()` where
`SUC` and `ZERO` are both `data X = MkX`.